posts/map_blocks_example.md

```diff
@@ -0,0 +1,191 @@
+---
+author: Matt Long
+date: 2021-4-22
+tags: xarray, dask, pop
+---
+
+# How to Use `xarray.map_blocks` for Vertical Interpolation of a 3D Field
+
+Within this example, we cover how to use `xarray.map_blocks` to calculate the mixed-layer depth within the CESM POP model output.
+
+This calculation is "embarassingly parallel" such that each calculation is done within a single a column. The calculation **_should_** be easily computed within each column across the model domain. This is where `map_blocks` can be used to improve the performance of this metric.
+
+## Imports and Data Ingestion
+
+We use the following libraries/packages within this example
+
+```python
+import os
+
+from scipy import interpolate
+import numpy as np
+import xarray as xr
+
+import matplotlib.pyplot as plt
+
+import pop_tools
+```
+
+Next, we read in the data using `xarray` and subset for temperature and salinity, subsetting for a smaller portion of the domain for testing purposes
+
+```
+ds = xr.merge((
+    xr.open_dataset('/glade/p/cgd/oce/projects/cesm2-marbl/xpersist_cache/3d_fields/TEMP-presentday-monclim.nc'),
+    xr.open_dataset('/glade/p/cgd/oce/projects/cesm2-marbl/xpersist_cache/3d_fields/SALT-presentday-monclim.nc'),
+))
+
+
+ds = ds[['TEMP', 'SALT']]
+
+# subset for testing purposes
+ds = ds.isel(member_id=slice(0, 2), nlat=slice(2, 5), nlon=slice(1, 5))
+```
+
+## Data Operation
+
+Next, we setup a function to compute the mixed layer depth (MLD). The MLD is defined as the point in the water column where the potential density (`sigma`) exceeds the surface density by a specified threshold (`dsigma`).
+
+```python
+def mld_dsigma(SALT, TEMP, dsigma=0.03, rho_chunks={'nlat': 16, 'nlon': 16}):
+    """
+    Compute MLD based on ∆σ criterion. Uses xarray.map_blocks.
+
+    Parameters
+    ----------
+
+    SALT : xarray.DataArray
+      Salinity
+    TEMP : xarray.DataArray
+      Potential temperature
+    dsigma : float, optional
+      The value for ∆σ.
+    rho_chunks: dictionary, optional
+      Dimension chunk parameters to chunk rho by
+
+    Returns
+    -------
+
+    MLD : xarray.DataArray
+      The MLD (m) defined as the point in the water column where
+      density exceeds rho[0] + dsigma.
+    """
+
+    # determine dimensionality
+    dims_in = SALT.dims
+    assert dims_in == TEMP.dims, 'dimension mismatch'
+    assert 'z_t' in SALT.coords, 'z_t not found in SALT coords'
+
+    # drop ancillary coordinates (this may not be necessary)
+    SALT = SALT.reset_coords(drop=True)
+    TEMP = TEMP.reset_coords(drop=True)
+
+    # compute density
+    rho = pop_tools.eos(SALT.chunk({'z_t': 10}),
+                        TEMP.chunk({'z_t': 10}),
+                        depth=SALT.z_t * 0.).compute()
```

Comment:
`z_t` is a "core dimension" for this calculation. You can't avoid having it be single-chunked in this case. Well you can by computing :P but that isn't right...

```suggestion
    assert len(SALT.chunks["z_t"]) == 1
    assert len(TEMP.chunks["z_t"]) == 1
    rho = pop_tools.eos(SALT, TEMP, depth=SALT.z_t * 0.)
```

Reply:
I am not sure I understand this entirely. Are you referring to a limitation in the `apply_ufunc` call in `eos.py`:
https://github.com/NCAR/pop-tools/blob/9563759ab063ae51356af18eab6cabe901e907bc/pop_tools/eos.py#L116
?

Also, I found that the `.compute` on `rho` seemed to improve performance, perhaps by truncating the task graph. This may be taken care of by the `persist` call below.

Reply:
I think my lowest order question is: why is this function not entirely lazy? Isn't `rho` potentially quite large; in which case the `compute` here would break things.

I think it should be lazy; in which case the only way the MLD calculation can work is if there is a single chunk along `z_t`. Hence my suggestion to ensure that.

Reply:
If rho is very large the `compute` will break things. In my case, `rho` fit in memory, so this didn't break things. However, I think more testing is needed as I subsequently added the `.persist` call below. I found that at least one of these (`compute` or `persist`) was necessary, otherwise the calculation just bogged down with too many tasks—but it might be that `persist` is all that's needed.

To be clear, for the MLD calculation, all dimensions *but* `z_t` can be chunked. However, if these all have a chunk size of one, there are too many tasks and it doesn't work.

Reply:
> n my case, rho fit in memory, so this didn't break things. 

We should at least describe these tradeoffs. I get the sense that people are looking at this blog as a kind of "best practices" recommendation, so we should be careful to do that. So I would still prefer a fully lazy implementation followed by a discussion of when things might break.

> To be clear, for the MLD calculation, all dimensions but z_t can be chunked. However, if these all have a chunk size of one, there are too many tasks and it doesn't work.

Yes this is what I was trying to say in a not very clear way. The proposed assert statements enforces this requirement on the inputs: `SALT`, `TEMP`. So then it is satisfied on `rho`

The "too many tasks" issue is a byproduct of using `map_blocks`. That's just how it works. It should ultimately benefit from ongoing dask work to make dealing with large graphs easier.

Reply:
> The "too many tasks" issue is a byproduct of using map_blocks. That's just how it works. It should ultimately benefit from ongoing dask work to make dealing with large graphs easier.

I think I managed to set this up so that the interpolation kernel operates on chunks with `len(non_vertical_dims) > 1`—so the parallelization is not across every `i, j` column (for example), but over blocks of data chunked so as to yield a reasonable number of tasks and appropriate amortization of chunk memory overhead across substantial compute.